specifyweb/frontend/js_src/lib/components/Statistics/hooks.tsx

```diff
@@ -334,53 +286,136 @@ export function useStatValueLoad<
   backEndResponse: BackendStatsResult | undefined,
   handleChange: (
     newCategories: (
-      oldCategory: StatLayout[number]['categories']
-    ) => StatLayout[number]['categories']
+      oldCategory: StatLayout['categories']
+    ) => StatLayout['categories']
   ) => void,
-  statsSpec: StatsSpec
+  categoriesToFetch: RA<string>
 ) {
-  React.useLayoutEffect(() => {
-    Object.entries(statsSpec).forEach(([pageName, pageSpec]) =>
-      Object.entries(pageSpec).forEach(([categoryName, categorySpec]) =>
-        Object.entries(categorySpec.items ?? {}).forEach(
-          ([itemName, { spec }]) => {
-            if (
-              itemName === 'phantomItem' &&
-              spec.type === 'BackEndStat' &&
-              Object.keys(backEndResponse ?? {}).includes(categoryName)
-            ) {
-              handleChange((oldCategory) =>
-                oldCategory.map((unknownCategory) => ({
-                  ...unknownCategory,
-                  items:
-                    unknownCategory.items ??
-                    (unknownCategory.categoryToFetch === undefined ||
-                    backEndResponse?.[unknownCategory.categoryToFetch] ===
-                      undefined ||
-                    unknownCategory.categoryToFetch !== categoryName
-                      ? undefined
-                      : Object.entries(backEndResponse[categoryName]).map(
-                          ([itemName, rawValue]) => ({
-                            type: 'DefaultStat',
-                            pageName,
-                            itemName: 'phantomItem',
-                            categoryName,
-                            label: itemName,
-                            itemValue: spec.formatter(rawValue),
-                            itemType: 'BackEndStat',
-                            pathToValue: itemName,
-                          })
-                        )),
-                }))
-              );
-            }
-          }
-        )
-      )
-    );
-  }, [backEndResponse, handleChange, statsSpec]);
+  React.useEffect(() => {
+    dynamicStatsSpec.forEach(({ responseKey, formatter }) => {
+      if (
+        backEndResponse !== undefined &&
+        backEndResponse[responseKey] !== undefined &&
+        categoriesToFetch.includes(responseKey)
+      ) {
+        handleChange((oldCategory) =>
+          oldCategory.map((unknownCategory) => ({
+            ...unknownCategory,
+            items: applyStatBackendResponse(
+              backEndResponse,
+              unknownCategory.items,
+              responseKey,
+              formatter,
+              statsSpec
+            ),
+          }))
+        );
+      }
+    });
+  }, [backEndResponse, handleChange]);
+}
+
+export function generateStatUrl(
+  urlPrefix: string,
+  categoryKey: string,
+  itemKey: string
+) {
+  const urlSpec = [urlPrefix, categoryKey, itemKey];
+  return `/statistics${urlSpec.reduce(
+    (previousValue, currentValue) =>
```

Comment:
Can this be replaced with `.filter()` and `.join()` to produce a much more readable code?